lib/charms/vault_k8s/v0/vault_kv.py

```diff
@@ -0,0 +1,369 @@
+        self.relation_id = snapshot["relation_id"]
+        self.relation_name = snapshot["relation_name"]
+        self.secret_backend = snapshot["secret_backend"]
+
+
+class DepartedVaultKvClientsEvent(ops.EventBase):
+    """Departed VaultKvClients Event."""
+
+    pass
+
+
+class GoneAwayVaultKvClientsEvent(ops.EventBase):
+    """GoneAway VaultKvClients Event."""
+
+    pass
+
+
+class VaultKvProviderEvents(ops.ObjectEvents):
+    """List of events that the Vault Kv provider charm can leverage."""
+
+    has_vault_kv_clients = ops.EventSource(HasVaultKvClientsEvent)
+    ready_vault_kv_clients = ops.EventSource(ReadyVaultKvClientsEvent)
+    departed_vault_kv_clients = ops.EventSource(DepartedVaultKvClientsEvent)
+    gone_away_vault_kv_clients = ops.EventSource(GoneAwayVaultKvClientsEvent)
+
+
+class VaultKvProvides(ops.Object):
+    """Class to be instanciated by the providing side of the relation."""
+
+    on = VaultKvProviderEvents()
+    _stored = ops.StoredState()
+
+    def __init__(
+        self,
+        charm: ops.CharmBase,
+        relation_name: str,
+    ) -> None:
+        super().__init__(charm, relation_name)
+        self.charm = charm
+        self.relation_name = relation_name
+        self.framework.observe(
+            self.charm.on[relation_name].relation_joined,
+            self._on_vault_kv_relation_joined,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_changed,
+            self._on_vault_kv_relation_changed,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_departed,
+            self._on_vault_kv_relation_departed,
+        )
+        self.framework.observe(
+            self.charm.on[relation_name].relation_broken,
+            self._on_vault_kv_relation_broken,
+        )
+
+    def _on_vault_kv_relation_joined(self, event: ops.RelationJoinedEvent):
+        """Handle client joined relation.
+
+        Args:
+            event: The event that triggered the handler.
+        """
+        self.on.has_vault_kv_clients.emit()
+
+    def _on_vault_kv_relation_changed(self, event: ops.RelationChangedEvent):
+        """Handle client changed relation."""
+        if event.app is None:
+            logger.debug("No remote application yet")
+            return
+
+        secret_backend = event.relation.data[event.app].get("secret_backend")
+
+        if secret_backend is not None:
+            self.on.ready_vault_kv_clients.emit(
+                event.relation.id,
+                event.relation.name,
+                secret_backend,
+            )
+
+    def _on_vault_kv_relation_departed(self, event: ops.RelationDepartedEvent):
+        """Handle client departed relation."""
+        self.on.departed_vault_kv_clients.emit()
+
+    def _on_vault_kv_relation_broken(self, event: ops.RelationBrokenEvent):
+        """Handle client broken relation."""
+        self.on.gone_away_vault_kv_clients.emit()
```

Comment:
Not sure about the value of this event, why wrap the relation broken and give it a new unfamiliar name at all?